src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -665,6 +740,67 @@ private static bool IsApplicableTriggerCharacter(string triggerCharacter, RazorL
             return false;
         }
 
+        private static string GetIndentationString(int indentation, FormattingOptions options)
+        {
+            if (options.InsertSpaces)
+            {
+                return new string(' ', indentation);
+            }
+            else
+            {
+                var tabs = indentation / options.TabSize;
+                var tabPrefix = new string('\t', (int)tabs);
+
+                var spaces = indentation % options.TabSize;
+                var spaceSuffix = new string(' ', (int)spaces);
+
+                var combined = string.Concat(tabPrefix, spaceSuffix);
+                return combined;
+            }
+        }
+
+        private bool IsWordOnEmptyLine(TextExtent? wordExtent, LSPDocumentSnapshot documentSnapshot)
+        {
+            if (wordExtent == null)
+            {
+                return false;
+            }
+
+            var line = wordExtent.Value.Span.Start.GetContainingLine();
+            var lineStart = line.Start.Position;
+            var wordStart = wordExtent.Value.Span.Start.Position;
+
+            // Is the word prefixed by whitespace?
+            for (var i = lineStart; i < wordStart; i++)
+            {
+                if (!char.IsWhiteSpace(documentSnapshot.Snapshot[i]))
+                {
+                    return false;
+                }
+            }
+
+            var lineEnd = line.End.Position;
+            var wordEnd = wordExtent.Value.Span.End.Position;
+
+            // Is the word suffixed by whitespace?
+            for (var i = wordEnd; i < lineEnd; i++)
+            {
+                if (!char.IsWhiteSpace(documentSnapshot.Snapshot[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static int GetBaseIndentation(TextExtent? wordExtent)
+        {
+            var lineStart = wordExtent.Value.Span.Start.GetContainingLine().Start.Position;
+            var baseIndentation = wordExtent.Value.Span.Start.Position - lineStart;
```

Comment:
What is indentation here? Levels, or characters? Because if its characters this calculation will be incorrect if the user is using tabs, right?

Reply:
I'm actually re-using our formatting logic to calculate a valid indentation level. Does that logic have a base assumption of 4 characters == 1 indentation?

Reply:
Fixed!